UIPS/UIP-0120.md

```diff
@@ -0,0 +1,79 @@
+---
+uip: "013X"
+title: Assertion Rune
+description: "Add a ~: sigcol rune to handle positive assertions cleanly"
+author: ~lagrev-nocfep (@sigilante)
+status: Draft
+type: Standards Track
+category: Hoon
+created: 2024-03-25
+---
+
+## Abstract
+
+Hoon supports several related runes for error handling and debugging:
+
+- `~&` is the classic `printf` side-effect rune.
+- `~|` adds a stack annotation in case of crash.
+- `~?` adds a debugging printf
+- `?>` is a positive assertion (crash if fail).
+- `?<` is a negative assertion (crash if pass).
+
+A common design pattern in Hoon is to have an assertion to succeed, then to wrap
+an intended error mesage with `~|` around `!!`.  By coupling the `?:`, `~|`, and
+`!!` together, we can achieve a more lapidary yet still clear code expression.
+We propose to call this the `~:` sigcol rune.
+
+```
+~:  test-to-pass
+  'message-on-fail'
+continuation if pass
+```
+
+## Specification
+
+The `~:` rune should desugar as follows:
+
+```
+~:  test-to-pass
+  'message-on-fail'
+continuation if pass
+```
+
+```
+?:  test-to-pass
```

Comment:
I think this should be a `?.`, not a `?:`

Reply:
er, yes